internal/controller/runtime/fsm/runtime_fsm_apply_clusterrolebindings.go

```diff
@@ -0,0 +1,192 @@
+	}
+)
+
+func sFnApplyClusterRoleBindings(ctx context.Context, m *fsm, s *systemState) (stateFn, *ctrl.Result, error) {
+	// prepare subresource client to request admin kubeconfig
+	srscClient := m.ShootClient.SubResource("adminkubeconfig")
+	shootAdminClient, err := GetShootClient(ctx, srscClient, s.shoot)
+	if err != nil {
+		updateCRBApplyFailed(&s.instance)
+		return updateStatusAndStopWithError(err)
+	}
+	// list existing cluster role bindings
+	var crbList rbacv1.ClusterRoleBindingList
+	if err := shootAdminClient.List(ctx, &crbList); err != nil {
+		updateCRBApplyFailed(&s.instance)
+		return updateStatusAndStopWithError(err)
+	}
+
+	removed := getRemoved(crbList.Items, s.instance.Spec.Security.Administrators)
+	missing := getMissing(crbList.Items, s.instance.Spec.Security.Administrators)
+
+	for _, fn := range []func() error{
+		newDelCRBs(ctx, shootAdminClient, removed),
+		newAddCRBs(ctx, shootAdminClient, missing),
+	} {
+		if err := fn(); err != nil {
+			updateCRBApplyFailed(&s.instance)
+			return updateStatusAndStopWithError(err)
+		}
+	}
+
+	s.instance.UpdateStateReady(
+		imv1.ConditionTypeRuntimeConfigured,
+		imv1.ConditionReasonConfigurationCompleted,
+		"kubeconfig admin access updated",
+	)
+	return updateStatusAndStop()
+}
+
+//nolint:gochecknoglobals
+var GetShootClient = func(ctx context.Context,
+	adminKubeconfigClient client.SubResourceClient, shoot *gardener_api.Shoot) (client.Client, error) {
+	// request for admin kubeconfig with low expiration timeout
+	var req authenticationv1alpha1.AdminKubeconfigRequest
+	if err := adminKubeconfigClient.Create(ctx, shoot, &req); err != nil {
+		return nil, err
+	}
+
+	restConfig, err := clientcmd.RESTConfigFromKubeConfig(req.Status.Kubeconfig)
+	if err != nil {
+		return nil, err
+	}
+
+	shootClientWithAdmin, err := client.New(restConfig, client.Options{})
+	if err != nil {
+		return nil, err
+	}
+
+	return shootClientWithAdmin, nil
+}
+
+func isRBACUserKind(s rbacv1.Subject) bool {
+	return s.Kind == rbacv1.UserKind &&
+		s.APIGroup == rbacv1.GroupName
+}
+
+func getRemoved(crbs []rbacv1.ClusterRoleBinding, admins []string) (removed []rbacv1.ClusterRoleBinding) {
+	// iterate over cluster role bindings to find out removed administrators
+	for _, crb := range crbs {
+		if !labels.Set(crb.Labels).AsSelector().Matches(labels.Set(labelsClusterRoleBindings)) {
+			// cluster role binding is not controlled by KIM
+			continue
+		}
+
+		index := slices.IndexFunc(crb.Subjects, isRBACUserKind)
+		if index < 0 {
+			// cluster role binding does not contain user subject
+			continue
+		}
+
+		subjectUserName := crb.Subjects[index].Name
+		if slices.Contains(admins, subjectUserName) {
+			continue
+		}
+		// administrator was removed
+		removed = append(removed, crb)
+	}
+
+	return removed
+}
+
+//nolint:gochecknoglobals
+var newContainsAdmin = func(admin string) func(rbacv1.ClusterRoleBinding) bool {
+	return func(r rbacv1.ClusterRoleBinding) bool {
+		for _, subject := range r.Subjects {
```

Comment:
You should also add some checking if given ClusterRole Binding has correct labels:

		if !labels.Set(r.Labels).AsSelector().Matches(labels.Set(labelsClusterRoleBindings)) {
			// cluster role binding is not controlled by KIM
			return false
		}

Reply:
We are checking those labels:
```
	labelsClusterRoleBindings = map[string]string{
		"app":                                   "kyma",
		"reconciler.kyma-project.io/managed-by": "kim",
	}
```